File: Sources/DoNotTypeCore/HistoryQuery.swift
import Foundation

/// Filtering and search over stored dictations.
///
/// Lives in the core rather than in each platform's list view so that "find that thing I dictated
/// last Tuesday" behaves identically everywhere, and so the matching rules can be tested without a
/// UI. Search is the point of keeping history at all — a log you cannot search is just disk usage.
public struct HistoryQuery: Sendable, Equatable {
    public enum StatusFilter: String, CaseIterable, Sendable {
        case all
        case completed
        case needsAttention

        public var label: String {
            switch self {
            case .all: "All"
            case .completed: "Completed"
            case .needsAttention: "Needs retry"
            }
        }
    }

    public var text: String = ""
    public var status: StatusFilter = .all
    /// Restrict to dictations made in one app.
    public var appName: String?
    public var since: Date?

    public init(
        text: String = "", status: StatusFilter = .all, appName: String? = nil, since: Date? = nil
    ) {
        self.text = text
        self.status = status
        self.appName = appName
        self.since = since
    }

    public var isEmpty: Bool {
        text.trimmed.isEmpty && status == .all && appName == nil && since == nil
    }

    /// Applies the filters, newest first.
    ///
    /// Matching is diacritic- and case-insensitive so searching "cafe" finds "café", and it
    /// searches the error text as well as the transcript — when you are looking for a failure,
    /// the message is the thing you remember.
    public func apply(to records: [DictationRecord]) -> [DictationRecord] {
        let needle = text.trimmed
        return records
            .filter { record in
                switch status {
                case .all: true
                case .completed: record.status == .completed
                case .needsAttention: record.status != .completed
                }
            }
            .filter { record in
                guard let appName else { return true }
                return record.appName == appName
            }
            .filter { record in
                guard let since else { return true }
                return record.createdAt >= since
            }
            .filter { record in
                guard !needle.isEmpty else { return true }
                return Self.matches(record: record, needle: needle)
            }
            .sorted { $0.createdAt > $1.createdAt }
    }

    static func matches(record: DictationRecord, needle: String) -> Bool {
        let haystacks = [
            record.text,
            record.errorMessage ?? "",
            record.appName ?? "",
            record.windowTitle ?? "",
        ]
        return haystacks.contains { $0.containsLoosely(needle) }
    }

    /// Apps that appear in the history, for populating a filter control.
    public static func appNames(in records: [DictationRecord]) -> [String] {
        Array(Set(records.compactMap(\.appName).filter { !$0.isEmpty })).sorted()
    }
}

extension String {
    /// Case- and diacritic-insensitive substring match.
    func containsLoosely(_ needle: String) -> Bool {
        guard !needle.isEmpty else { return true }
        return range(of: needle, options: [.caseInsensitive, .diacriticInsensitive]) != nil
    }
}

extension HistoryStore {
    /// Convenience so callers do not have to load and filter in two steps.
    public func search(_ query: HistoryQuery) -> [DictationRecord] {
        query.apply(to: all())
    }
}

File: windows/DoNotType.Core/HistoryQuery.cs
namespace DoNotType.Core;

/// <summary>
/// Filtering and search over stored dictations.
///
/// In the core rather than in the settings form so that "find that thing I dictated last Tuesday"
/// behaves the same on every platform, and so the matching rules can be tested without a UI.
/// Search is the point of keeping history at all — a log you cannot search is just disk usage.
/// </summary>
public sealed class HistoryQuery
{
    public enum StatusFilter
    {
        All,
        Completed,
        NeedsAttention,
    }

    public string Text { get; set; } = string.Empty;
    public StatusFilter Status { get; set; } = StatusFilter.All;
    public string? AppName { get; set; }
    public DateTimeOffset? Since { get; set; }

    public bool IsEmpty =>
        string.IsNullOrWhiteSpace(Text) && Status == StatusFilter.All
        && AppName is null && Since is null;

    /// <summary>Applies the filters, newest first.</summary>
    public IReadOnlyList<DictationRecord> Apply(IEnumerable<DictationRecord> records)
    {
        var needle = Text.Trim();

        return records
            .Where(r => Status switch
            {
                StatusFilter.Completed => r.Status == DictationStatus.Completed,
                StatusFilter.NeedsAttention => r.Status != DictationStatus.Completed,
                _ => true,
            })
            .Where(r => AppName is null || r.AppName == AppName)
            .Where(r => Since is null || r.CreatedAt >= Since)
            .Where(r => needle.Length == 0 || Matches(r, needle))
            .OrderByDescending(r => r.CreatedAt)
            .ToList();
    }

    /// <summary>
    /// Error text is searched as well as the transcript: when you are hunting a failure, the
    /// message is what you remember — the transcript is empty.
    /// </summary>
    internal static bool Matches(DictationRecord record, string needle) =>
        new[] { record.Text, record.ErrorMessage ?? string.Empty,
                record.AppName ?? string.Empty, record.WindowTitle ?? string.Empty }
            .Any(haystack => haystack.Contains(needle, StringComparison.OrdinalIgnoreCase));

    /// <summary>Apps present in the history, for populating a filter control.</summary>
    public static IReadOnlyList<string> AppNames(IEnumerable<DictationRecord> records) =>
        records.Select(r => r.AppName)
            .Where(name => !string.IsNullOrEmpty(name))
            .Select(name => name!)
            .Distinct()
            .OrderBy(name => name)
            .ToList();
}
